### PR TITLE
Memory reduction

### PR DESCRIPTION
**Pull request by drewnoakes**
Various reductions to memory, mainly avoiding temporary strings and arrays. I searched the project for `Trim` and `Split`.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs [hunk 1]
```diff
@@ -37,6 +37,25 @@ public LazyStringSplit(string input, char delimiter)
 
         IEnumerator<string> IEnumerable<string>.GetEnumerator() => GetEnumerator();
 
+        public IEnumerable<T> Select<T>(Func<string, T> func)
```
Comment (abpiskunov) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs]: Nice util, I wish this type of helpful utils were shared somewhere across VS for others to reuse ..

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Build/BuildUtilities.cs [hunk 2]
```diff
@@ -66,21 +66,16 @@ public static ProjectPropertyElement GetProperty(ProjectRootElement project, str
...
             {
-                if (!string.IsNullOrEmpty(value))
+                if (!string.IsNullOrWhiteSpace(value))
                 {
-                    valuesBuilder.Add(value);
+                    yield return value.Trim();
                 }
             }
-
-            return valuesBuilder.Distinct(StringComparer.OrdinalIgnoreCase).ToImmutableArray();
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Build/BuildUtilities.cs]: This change might not be correct actually. The previous code eliminated duplicates. Will review.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Configuration/BaseProjectConfigurationDimensionProvider.cs [hunk 3]
```diff
@@ -71,7 +72,7 @@ protected virtual async Task<ImmutableArray<string>> GetOrderedPropertyValuesAsy
             }
             else
             {
-                return BuildUtilities.GetPropertyValues(propertyValue);
+                return BuildUtilities.GetPropertyValues(propertyValue).ToImmutableArray();
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Configuration/BaseProjectConfigurationDimensionProvider.cs]: The change was sympathetic to other callers that could consume values lazily. Though that may not make sense if it turns out I reinstate de-duplicating the output of GetPropertyValues, in which case I may restore the immutable return type.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs]: Where would it live?

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs]: Where would it live?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Build/BuildUtilities.cs (hunk 2 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/Build/BuildUtilities.cs]: I restored deduplication and added a test with duplicates.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Configuration/BaseProjectConfigurationDimensionProvider.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Configuration/BaseProjectConfigurationDimensionProvider.cs]: A ImmutableArray builder versus the `ToImmutableArray` method are on par for perf. However for lazy consumers (there are a few), we avoid allocating more than just the enumerator, and if they don't consume all elements then we don't allocate substrings that would be unused.

**Pull request merged**